include/mqtt/endpoint.hpp

```diff
@@ -9160,56 +9160,50 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
             mqtt_connected_ = true;
             switch (version_) {
             case protocol_version::v3_1_1:
-                if (!on_connect(
+                if (on_connect(
                         force_move(info.client_id),
                         force_move(info.user_name),
                         force_move(info.password),
-                        [&] () -> optional<will> {
-                            if (connect_flags::has_will_flag(info.connect_flag)) {
-                                return
-                                    will(
-                                        force_move(info.will_topic),
-                                        force_move(info.will_payload),
-                                        connect_flags::has_will_retain(info.connect_flag),
-                                        connect_flags::will_qos(info.connect_flag)
-                                    );
-                            }
-                            return nullopt;
-                        } (),
+                          connect_flags::has_will_flag(info.connect_flag)
+                        ? optional<will>(
+                              will(
+                                   force_move(info.will_topic),
+                                   force_move(info.will_payload),
+                                   connect_flags::has_will_retain(info.connect_flag),
+                                   connect_flags::will_qos(info.connect_flag)
+                                  )
+                            )
+                        : optional<will>(nullopt),
```

Comment:
One question.

You use `?:` conditional operator instead of the lambda expression.
I think that lambda expression is better than conditional operator because the lambda expression can define return value type and easy to understand.

For example

```cpp
int a = b > c ? d
              : e > f ? g
                      : h;
```

is difficult to understand but

```cpp
int a = [&] {
    if (b > c) return d;
    else if (e > f) return g;
    else  h;
} ();
```

is easy.

I guess that for optional value, the lambda expression approach might be inefficient so you repleace it with `?:`.
Is that right? `-O3` optimization doesn't work for it?

Reply:
I found that using mqtt_cpp::optional with the lambda "immediately executed function expression"  was confusing.

Part of the issue with optional is that when not compiling on -O3 (which I don't do most of the time to help with debugging) it adds several additional function calls to the call stack, which makes things difficult to follow.

Do you have a strong preference for using the lambdas?

Maybe we can do something like MQTT_CONDITIONAL(condition, success, failure) macro?

Reply:
> Do you have a strong preference for using the lambdas?

No, I just want to ask why do you change it.
But I don't want to replace all other lambda immediate call.
Lambda expression is more flexible than `?:` for example lambda expression can use temporary variable, can define return value explicitly.

> Maybe we can do something like MQTT_CONDITIONAL(condition, success, failure) macro?

I guess that it is not temporary variable friendly.
The current master is mixed lambda immediate call and `?:`. I think that it is not a problem.

Reply:
Ok, that's fine.